### PR TITLE
Fixed up some code to generate question/answer pairs given a chunk.

### PR DESCRIPTION
**Pull request by ivanleomk**
Refactored out the chunking code into a `src/chunking` file and also added a simple instructor function to generate question/answer pairs given a specific chunk.

Problem seems to be that chunk doesn't have that much information if we just split by newlines. Most are just 1-2 sentences long.

Also added an example of output in `output.json`

----

| <a href="https://ellipsis.dev" target="_blank">Ellipsis</a> | :rocket: This PR description was created by [Ellipsis](https://www.ellipsis.dev) for commit 274d2654cd8f1168460ecf411a63c6b1e921bb9f.  | 
|--------|--------|

### Summary:
This PR refactors the chunking code, introduces a new function to generate question-answer pairs, updates the CLI to include these new features and an option to specify the output file path for the 'synthethic_questions' command, and assumes updates in `ingest.py` and `models.py` and new tests in `tests/test_models.py`.

**Key points**:
- Refactored chunking code into `src/chunking.py`
- Added `generate_question_answer_pair` function in `generate_synthetic_question.py`
- Updated CLI command 'synthethic_questions' to include an option for specifying the output file path
- Generated question-answer pairs are saved in `output.json` by default
- Assumed updates in `ingest.py` and `models.py` and new tests in `tests/test_models.py`

----
Generated with :heart: by [ellipsis.dev](https://www.ellipsis.dev)

### CODE REVIEW

**Comment by ivanleomk** [comment]
Might also convert the question generation to async to support more questions and also the prompts for the questions.

File: output.json [hunk 1]
```diff
@@ -0,0 +1,52 @@
+[
+  {
+    "question": "What aspect of performance returns did the author not understand as a child?",
+    "answer": "superlinear returns for performance",
+    "chunk": "One of the most important things I didn't understand about the world when I was a child is the degree to which the returns for performance are superlinear."
```
Comment (jxnl) [inline output.json]: should be chunk_id too since we want to check if its returned

File: rag_app/generate.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+    This model represents a pair of a question generated from a text chunk, its corresponding answer,
+    and the chain of thought leading to the answer. The chain of thought provides insight into how the answer
+    was derived from the question.
+    """
+
+    question: str = Field(
+        ..., description="The generated question from the text chunk."
+    )
+    answer: str = Field(..., description="The answer to the generated question.")
+    chain_of_thought: str = Field(
```
Comment (jxnl) [inline rag_app/generate.py]: chain_of_thought should be first!

File: rag_app/generate.py [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+    output = []
+    for idx, chunk in tqdm(enumerate(chunks), total=max_questions):
+        if idx == max_questions:
+            break
+        response = generate_question_answer_pair(chunk["text"])
+        output.append(
+            {
+                "question": response.question,
+                "answer": response.answer,
+                "chunk": chunk["text"],
```
Comment (jxnl) [inline rag_app/generate.py]: add chunk_id

File: rag_app/generate.py [hunk 4]
```diff
@@ -0,0 +1,74 @@
+import typer
```
Comment (jxnl) [inline rag_app/generate.py]: generate.py is too generic

`generate_synthetic_questions.py`

File: rag_app/generate.py [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+    file = read_files(Path(folder_path), file_suffix=".md")
+    chunks = chunk_text(file)
+
+    import json
+
+    output = []
+    for idx, chunk in tqdm(enumerate(chunks), total=max_questions):
+        if idx == max_questions:
+            break
+        response = generate_question_answer_pair(chunk["text"])
```
Comment (jxnl) [inline rag_app/generate.py]: use asyncio, too slow

**Comment by ivanleomk** [comment]
> Found a bug in the code, will push up a fix later. Please don't merge in yet

This has been fixed, the `query` and `generate` functions work fine now

File: rag_app/generate_synthetic_question.py [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+        ],
+        response_model=QuestionAnswerPair,
+    )
+    return (res, chunk)
+
+
+async def gather_questions(chunks) -> List[tuple[QuestionAnswerPair, TextChunk]]:
+    coros = [generate_question_answer_pair(chunk) for chunk in chunks]
+    output = []
+    for response in tqdm.asyncio.tqdm_asyncio.as_completed(coros):
```
Comment (jxnl) [inline rag_app/generate_synthetic_question.py]: do 

`from tqdm.asyncio import tdqm_asyncio as asyncio`

that way its always asyncio.as_completed, so if we remove it we dont have to edit code.

Comment (ivanleomk) [reply rag_app/generate_synthetic_question.py]: fixed this up

**Pull request merged**